content/docs/specs/native_histograms.md

```diff
@@ -0,0 +1,2416 @@
+observations, for example to count the number of seconds an observed value was
+falling into different buckets of a histogram. The far more common use case for
+float histograms is within PromQL, though. PromQL generally only acts on float
+values, so the PromQL engine converts every histogram retrieved from the TSDB
+to a float histogram first, and any histogram stored back into TSDB via
+recording rules is a float histogram. If such a histogram is effectively an
+integer histogram (because the value of all non-_sum_ fields can be represented
+precisely as uint64), a TSDB implementation MAY convert them back to integer
+histograms to increase storage efficiency. (As of Prometheus v3.00, the TSDB
+implementation within Prometheus is not utilizing this option.) Note, however,
+that the most common PromQL function applied to a counter histogram is `rate`,
+which generally produces non-integer numbers, so that results of recording
+rules will commonly be float histograms with non-integer values anyway.
+
+Treating native histograms explicitly as integer histograms vs. float histogram
+is a notable deviation from the treatment of conventional simple numeric
+samples, which are always treated as floats throughout the whole stack for the
+sake of simplicity.
+
+The main reason for the more involved treatment of histograms is the easy
+efficiency gains in protobuf-based exposition formats. Protobuf uses varint
+encoding for integers, which reduces the data size for small integer values
+without requiring an additional compression layer. This benefit is amplified by
+the [delta encoding of integer buckets](#buckets), which generally results in
+smaller integer values. Floats, in contrast, always require 8 bytes in
+protobuf. In practice, many integers in an integer histogram will fit in 1
+byte, and most will fit in 2 bytes, so that the explicit presence of integer
+histogram in a protobuf-exposition format results directly in a data size
+reduction approaching 8x for histograms with many buckets. This is particularly
+relevant as the overwhelming majority of histograms exposed by instrumented
+targets are integer histograms.
+
+For similar reasons, the representation of integer histograms in RAM and on
+disk is generally more efficient than that of float histograms. This is less
+relevant than the benefits in the exposition format, though. For one,
+Prometheus uses Gorilla-style XOR encoding for floats, which reduces their
+size, albeit not as much as the double-delta encoding used for integers. More
+importantly, an implementation could always decide to internally use an integer
+representation for histogram fields that are effectively integer values (see
+above). (Historical note: Prometheus v1 used exactly this approach to improve
+the compression of float samples, and Prometheus v3 might very well adopt this
+approach again in the future.)
+
+In a counter histogram, the total _count_ of observation and the counts in the
```

Comment:
I would expect this section to tell us what to do for gauge histograms. Are there representable by native histograms? If yes, then how?

Reply:
The concept of a gauge histogram is not linked to NH. It was explained (and referenced) above. There, it is explicitly said that a NH can have gauge flavor. Not sure what else there is to explain for the representation. (Should we say somewhere that in a gauge histogram, all the counts are gauges? That's kind of the idea of a gauge histogram and not specific to native histograms, so maybe we shouldn't create the impression it is. Note that this whole paragraph here exists because classic histograms were treating _sum_ wrongly from the beginning, and it got finally fixed with NH.)